### PR TITLE
Change categories mapping logic

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/data/pre_filtering.py [hunk 1]
```diff
@@ -88,7 +88,15 @@ def remove_unused_labels(dataset: DmDataset, data_format: str, ignore_index: int
         raise ValueError(msg)
     if len(used_labels) == len(original_categories):
         return dataset
+    if data_format == "arrow" and sorted(used_labels)[-1] != len(original_categories) - 1:
```
Comment (wonjuleee) [inline src/otx/core/data/pre_filtering.py:91]: Is it right to compare the empty label with the length of original categories?

Comment (eunwoosh) [inline src/otx/core/data/pre_filtering.py]:
```suggestion
    if data_format == "arrow" and max(used_labels) != len(original_categories) - 1:
```

**Review by eunwoosh — COMMENTED** [review]
Sorry but I can't understand what it's for. Could you explain more?

**Comment by kprokofi** [comment]
> Sorry but I can't understand what it's for. Could you explain more?

We now have a logic that removes unused labels from dataset. Sometimes it can happen that this pre-filtering removes unannotated objects from Geti:
image
It is undesired behavior that leads to 0 accuracy and issues. Unfortunately, Datumaro returns only IDs, not names. It is hard to distinguish empty label. But I see, that empty label is always in separated label group and it should be the last one. What I did - it is a temporary workaround.

File: src/otx/core/data/pre_filtering.py [hunk 2]
```diff
@@ -88,7 +88,15 @@ def remove_unused_labels(dataset: DmDataset, data_format: str, ignore_index: int
         raise ValueError(msg)
     if len(used_labels) == len(original_categories):
         return dataset
+    if data_format == "arrow" and max(used_labels) != len(original_categories) - 1:
+        # we assume that empty label is always the last one. If it is not explicitly added to the dataset,
+        # (not in the used labels) it will be filtered out.
```
Comment (harimkang) [inline src/otx/core/data/pre_filtering.py:93]: Can we verify that this assumption is always correct? Is there any code or documentation can reference for this?

Comment (kprokofi) [reply src/otx/core/data/pre_filtering.py:93]: I will try to find it on the Datumaro side

**Comment by eunwoosh** [comment]
@kprokofi , thanks for your kind explanation. Could you add TODO docstring?

File: src/otx/core/data/pre_filtering.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/core/data/pre_filtering.py:93]: Unfortunately, I haven't found the 100% proof for that. I can only see that labels are sorted in label schema by assigned ID
https://github.com/intel-innersource/frameworks.ai.interactive-ai-workflow.sonoma-creek/blob/386159bd64df784985efb2cc460d9a4b00d63435/src/sc_sdk/sc_sdk/otx/api/entities/label_schema.py#LL314
 And Empty label should have the largest ID value.

Comment (harimkang) [reply src/otx/core/data/pre_filtering.py:93]: @wonjuleee Can you confirm that we can use this workaround?

**Review by harimkang — APPROVED** [review]

**Review by chuneuny-emily — APPROVED** [review]

**Pull request merged**